### PR TITLE
Add dll injection to WindowsUtils

### PR DESCRIPTION
**Pull request by pierricgimmig**

### CODE REVIEW

File: src/WindowsUtils/Injection.cpp [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+  std::vector<Module> result;
+  for (const Module& module : modules) {
+    if (absl::EqualsIgnoreCase(module.name, module_name)) {
+      result.push_back(module);
+    }
+  }
+  if (result.size() == 1) return result[0];
+  if (result.size() > 1)
+    return ErrorMessage(
+        absl::StrFormat("Multiple modules with the name \"%s\" found", module_name));
```
Comment (dpallotti) [inline src/WindowsUtils/Injection.cpp:123]: This needs `{}`.

File: src/WindowsUtils/Injection.cpp [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+  return ErrorMessage("Could not find module in target process");
+}
+
+ErrorMessageOr<Module> FindModule(uint32_t pid, std::string_view module_name,
```
Comment (dpallotti) [inline src/WindowsUtils/Injection.cpp:127]: This method retries with sleeps so the name should be expanded. Something like `FindModuleWithRetries`.

File: src/WindowsUtils/Injection.cpp [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    if (result.has_value() || num_retries-- <= 0) return result;
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  }
+}
+
+ErrorMessageOr<void> EnsureModuleIsNotAlreadyLoaded(uint32_t pid, std::string_view module_name) {
+  ErrorMessageOr<Module> module = FindModule(pid, module_name);
+  if (module.has_value()) {
+    return ErrorMessage(
+        absl::StrFormat("Module %s is already loaded in process %u", module_name, pid));
```
Comment (dpallotti) [inline src/WindowsUtils/Injection.cpp:140]:
```suggestion
        absl::StrFormat("Module \"%s\" is already loaded in process %u", module_name, pid));
```

File: src/WindowsUtils/Injection.cpp [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+  SCOPED_TIMED_LOG("Injecting dll \"%s\" in process %u", dll_name, pid);
+
+  OUTCOME_TRY(ValidatePath(dll_path));
+  OUTCOME_TRY(EnsureModuleIsNotAlreadyLoaded(pid, dll_path.filename().string()));
+
+  // Inject dll by calling "LoadLibraryA" in remote process with the name of our dll as parameter.
+  OUTCOME_TRY(CreateRemoteThread(pid, "kernel32.dll", "LoadLibraryA", ToByteBuffer(dll_name)));
+
+  // Find injected dll in target process. Allow for retries as the loading might take some time.
+  OUTCOME_TRY(Module module, FindModule(pid, dll_path.filename().string(), /*num_retries=*/10));
```
Comment (dpallotti) [inline src/WindowsUtils/Injection.cpp:158]: Prefer a constant for `num_retries`

File: src/WindowsUtils/Injection.cpp [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+  LOG("Module %s successfully injected in process %u", dll_name, pid);
```
Comment (dpallotti) [inline src/WindowsUtils/Injection.cpp:160]:
```suggestion
  LOG("Module \"%s\" successfully injected in process %u", dll_name, pid);
```

File: src/WindowsUtils/Injection.cpp [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+ErrorMessageOr<void> CreateRemoteThread(uint32_t pid, std::string_view module_name,
+                                        std::string_view function_name,
+                                        std::vector<uint8_t> parameter) {
+  OUTCOME_TRY(uint64_t function_address, GetRemoteProcAddress(pid, module_name, function_name));
+  OUTCOME_TRY(HANDLE handle, GetHandleFromPid(pid));
+  orbit_base::unique_resource handle_closer(handle, ::CloseHandle);
+
+  // Write parameter to remote process memory.
+  void* parameter_address = nullptr;
```
Comment (dpallotti) [inline src/WindowsUtils/Injection.cpp:172]: I wrote this above, but again I don't think this should be a pointer.

File: src/WindowsUtils/Injection.cpp [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+                            parameter_address,
+                            /*dwCreationFlags=*/0, /*lpThreadId=*/0)) {
+    return ErrorMessage(absl::StrFormat("CreateRemoteThread failed with error %s",
+                                        orbit_base::GetLastErrorAsString()));
+  }
+
+  return outcome::success();
+}
+
+ErrorMessageOr<uint64_t> GetRemoteProcAddress(uint32_t pid, std::string_view module_name,
```
Comment (dpallotti) [inline src/WindowsUtils/Injection.cpp:189]: I think that this functions really needs some documenting comments. For example explaining some terminology. It's very hard to follow.

File: src/WindowsUtils/CMakeLists.txt [hunk 8]
```diff
@@ -22,6 +22,7 @@ target_include_directories(WindowsUtils PRIVATE
 target_sources(WindowsUtils PUBLIC
         include/WindowsUtils/AdjustTokenPrivilege.h
         include/WindowsUtils/FindDebugSymbols.h
+        include/WindowsUtils/Injection.h
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/WindowsUtils/CMakeLists.txt]: I agree, done, thanks.

File: src/WindowsUtils/InjectionTest.cpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace orbit_windows_utils {
+
+namespace {
+
+[[nodiscard]] std::filesystem::path GetTestDllPath() {
+  return orbit_test::GetTestdataDir() / "libtest.dll";
+}
+
+[[nodiscard]] std::filesystem::path GetNonExistantDllPath() { return "Z:/non_existant_dll.dll"; }
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/WindowsUtils/InjectionTest.cpp]: Done, thanks.

File: src/WindowsUtils/InjectionTest.cpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+}  // namespace
+
+TEST(Injection, InjectDlInCurrentProcess) {
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/WindowsUtils/InjectionTest.cpp]: Done, thanks.

File: src/WindowsUtils/InjectionTest.cpp [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+      GetRemoteProcAddress(pid, "libtest.dll", "PrintHelloWorld");
+  EXPECT_TRUE(remote_proc_result.has_value());
+
+  // CreateRemoteThread.
+  ErrorMessageOr<void> remote_thread_result =
+      CreateRemoteThread(pid, "libtest.dll", "PrintHelloWorld", {});
+  EXPECT_FALSE(remote_thread_result.has_error());
+}
+
+TEST(Injection, InjectNonExistantDll) {
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/WindowsUtils/InjectionTest.cpp]: Done, thanks.

File: src/WindowsUtils/Injection.cpp (hunk 3 above)
Comment (pierricgimmig) [reply src/WindowsUtils/Injection.cpp]: Done, thanks.

File: src/WindowsUtils/Injection.cpp (hunk 4 above)
Comment (pierricgimmig) [reply src/WindowsUtils/Injection.cpp]: Done, thanks.

File: src/WindowsUtils/Injection.cpp (hunk 5 above)
Comment (pierricgimmig) [reply src/WindowsUtils/Injection.cpp]: Done, thanks.

File: src/WindowsUtils/Injection.cpp (hunk 6 above)
Comment (pierricgimmig) [reply src/WindowsUtils/Injection.cpp]: I switch to uint64_t. Note that I didn't see the comment you are referring to, but I'm assuming it was concerning the fact that we had a `void*` that comes from a different address space (?).

File: src/WindowsUtils/Injection.cpp [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+ErrorMessageOr<Module> FindModule(uint32_t pid, std::string_view module_name,
+                                  uint32_t num_retries) {
+  while (true) {
+    ErrorMessageOr<Module> result = FindModule(pid, module_name);
+    if (result.has_value() || num_retries-- <= 0) return result;
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  }
+}
+
+ErrorMessageOr<void> EnsureModuleIsNotAlreadyLoaded(uint32_t pid, std::string_view module_name) {
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/WindowsUtils/Injection.cpp]: The ErrorMessageOr class is nodiscard, I prefer omitting [[nodiscard]] in this case.

File: src/WindowsUtils/Injection.cpp [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+// clang-format off
+#include <windows.h>
+#include <processthreadsapi.h>
+// clang-format on
+
+namespace orbit_windows_utils {
+
+namespace {
+
+ErrorMessageOr<void*> RemoteWrite(HANDLE process_handle, const std::vector<uint8_t>& buffer) {
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/WindowsUtils/Injection.cpp]: Yeah good point. I agree, let's move them when he have to.

File: src/WindowsUtils/include/WindowsUtils/Injection.h [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+#include <vector>
+
+#include "OrbitBase/Result.h"
+
+namespace orbit_windows_utils {
+
+// Injects a dll in a remote process identified by "pid".
+[[nodiscard]] ErrorMessageOr<void> InjectDll(uint32_t pid, std::filesystem::path dll_path);
+
+// Create a thread in a remote process and call specified function with "parameter" as argument.
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/WindowsUtils/include/WindowsUtils/Injection.h]: Added comment and link.

**Comment by pierricgimmig** [comment]
@dpallotti, @danielfenner thanks for the review! I addressed all the comments, please take another look, thanks.

File: src/WindowsUtils/CMakeLists.txt [hunk 15]
```diff
@@ -22,6 +22,7 @@ target_include_directories(WindowsUtils PRIVATE
 target_sources(WindowsUtils PUBLIC
         include/WindowsUtils/AdjustTokenPrivilege.h
         include/WindowsUtils/FindDebugSymbols.h
+        include/WindowsUtils/DllInjection.h
```
Comment (dpallotti) [inline src/WindowsUtils/CMakeLists.txt:25]: Keep alphabetical order.

File: src/WindowsUtils/CMakeLists.txt [hunk 16]
```diff
@@ -31,6 +32,7 @@ target_sources(WindowsUtils PUBLIC
 target_sources(WindowsUtils PRIVATE
         AdjustTokenPrivilege.cpp
         FindDebugSymbols.cpp
+        DllInjection.cpp
```
Comment (dpallotti) [inline src/WindowsUtils/CMakeLists.txt:35]: Alphabetical order

File: src/WindowsUtils/CMakeLists.txt [hunk 17]
```diff
@@ -45,6 +47,7 @@ add_executable(WindowsUtilsTests)
 
 target_sources(WindowsUtilsTests PRIVATE
         FindDebugSymbolsTest.cpp
+        DllInjectionTest.cpp
```
Comment (dpallotti) [inline src/WindowsUtils/CMakeLists.txt:50]: Alphabetical order

File: src/WindowsUtils/DllInjection.cpp [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+
+  if (num_bytes_written != buffer.size()) {
+    return ErrorMessage("WriteProcessMemory could not write the requested number bytes.");
+  }
+
+  return absl::bit_cast<uint64_t>(base_address);
+}
+
+template <typename T>
+ErrorMessageOr<void> RemoteRead(HANDLE process_handle, uint64_t base_address, T& result) {
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:50]: Why not return `ErrorMessageOr<T>`? Is it a lot of data and can the compiler not optimize? `ErrorMessageOr<std::string>` from `ReadRemoteString` seems cleaner.

File: src/WindowsUtils/DllInjection.cpp [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+  size_t number_of_bytes_read;
+  if (!ReadProcessMemory(process_handle, absl::bit_cast<LPCVOID>(base_address), &result, sizeof(T),
+                         &number_of_bytes_read)) {
+    return ErrorMessage(
+        absl::StrFormat("Error calling ReadProcessMemory: %s", orbit_base::GetLastErrorAsString()));
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:55]: We should be consistent with error messages. Here and below you write "Error calling ...: %s", while above you wrote "... failed with error %s".

File: src/WindowsUtils/DllInjection.cpp [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+  if (number_of_bytes_read != sizeof(T)) {
+    return ErrorMessage("Number of bytes read is not equal to requested size");
+  }
+
+  return outcome::success();
+}
+
+ErrorMessageOr<std::string> RemoteReadString(HANDLE process_handle, uint64_t base_address) {
+  std::string result;
+  char current_char = 'c';
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:67]: Why `'c'` specifically?

File: src/WindowsUtils/DllInjection.cpp [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+  size_t number_of_bytes_read = 0;
+  constexpr size_t read_size = sizeof(current_char);
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:69]: `kReadSize`

File: src/WindowsUtils/DllInjection.cpp [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+  std::vector<Module> result;
+  for (const Module& module : modules) {
+    if (absl::EqualsIgnoreCase(module.name, module_name)) {
+      result.push_back(module);
+    }
+  }
+  if (result.size() == 1) return result[0];
+  if (result.size() > 1)
+    return ErrorMessage(
+        absl::StrFormat("Multiple modules with the name \"%s\" found", module_name));
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:123]: This needs `{}`.

File: src/WindowsUtils/DllInjection.cpp [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+  return ErrorMessage("Could not find module in target process");
+}
+
+ErrorMessageOr<Module> FindModule(uint32_t pid, std::string_view module_name,
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:127]: This method retries with sleeps so the name should be expanded. Something like `FindModuleWithRetries`.

File: src/WindowsUtils/DllInjectionTest.cpp [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+      GetRemoteProcAddress(pid, "libtest.dll", "PrintHelloWorld");
+  EXPECT_TRUE(remote_proc_result.has_value());
+
+  // CreateRemoteThread.
+  ErrorMessageOr<void> remote_thread_result =
+      CreateRemoteThread(pid, "libtest.dll", "PrintHelloWorld", {});
+  EXPECT_FALSE(remote_thread_result.has_error());
+}
+
+TEST(Injection, InjectNonExistantDll) {
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjectionTest.cpp:52]:
```suggestion
TEST(Injection, InjectNonExistentDll) {
```

File: src/WindowsUtils/DllInjectionTest.cpp [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+[[nodiscard]] std::filesystem::path GetNonExistentDllPath() { return "Z:/non_existent_dll.dll"; }
+
+}  // namespace
+
+TEST(Injection, InjectDllInCurrentProcess) {
+  // Injection.
+  uint32_t pid = orbit_base::GetCurrentProcessId();
+  ErrorMessageOr<void> result = InjectDll(pid, GetTestDllPath());
+  if (result.has_error()) ERROR("Running InjectDllInCurrentProcess: %s", result.error().message());
+  EXPECT_FALSE(result.has_error());
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjectionTest.cpp:34]: You should use `EXPECT_THAT(...result, orbit_test_utils::HasNoError())`. That way, in case there is an error, it prints it, without the need to do it manually.

File: src/WindowsUtils/DllInjectionTest.cpp [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+
+  // Re-injection.
+  result = InjectDll(pid, GetTestDllPath());
+  EXPECT_TRUE(result.has_error());
+  EXPECT_THAT(result.error().message(), testing::HasSubstr("is already loaded in process"));
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjectionTest.cpp:39]: We have `orbit_test_utils::HasError`, it prints better messages in case of failure:
`EXPECT_THAT(result, HasError("Path does not exist"));`

File: src/WindowsUtils/DllInjectionTest.cpp [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+
+  // GetRemoteProcAddress.
+  ErrorMessageOr<uint64_t> remote_proc_result =
+      GetRemoteProcAddress(pid, "libtest.dll", "PrintHelloWorld");
+  EXPECT_TRUE(remote_proc_result.has_value());
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjectionTest.cpp:44]: `HasNoError`

File: src/WindowsUtils/DllInjectionTest.cpp [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+
+  // CreateRemoteThread.
+  ErrorMessageOr<void> remote_thread_result =
+      CreateRemoteThread(pid, "libtest.dll", "PrintHelloWorld", {});
+  EXPECT_FALSE(remote_thread_result.has_error());
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjectionTest.cpp:49]: `HasNoError`

File: src/WindowsUtils/DllInjectionTest.cpp [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+TEST(Injection, InjectNonExistantDll) {
+  uint32_t pid = orbit_base::GetCurrentProcessId();
+  ErrorMessageOr<void> result = InjectDll(pid, GetNonExistentDllPath());
+  EXPECT_TRUE(result.has_error());
+  EXPECT_THAT(result.error().message(), testing::HasSubstr("Path does not exist"));
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjectionTest.cpp:56]: `HasError`

File: src/WindowsUtils/include/WindowsUtils/DllInjection.h [hunk 30]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef WINDOWS_UTILS_INJECTION_H_
+#define WINDOWS_UTILS_INJECTION_H_
```
Comment (dpallotti) [inline src/WindowsUtils/include/WindowsUtils/DllInjection.h:6]: Adjust include guards.

File: src/WindowsUtils/include/WindowsUtils/Injection.h [hunk 31]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
```
Comment (dpallotti) [inline src/WindowsUtils/include/WindowsUtils/Injection.h:1]: Is this file duplicated now? Or is it again weird GitHub behaviour?

**Review by dpallotti — COMMENTED** [review]
Something weird happened in this pull request. As you said, you didn't see a comment. Well, multiple comments are only visible to me, but they are not pending. I can see they don't show up if I open this page in incognito mode. Some GitHub bug. I added them again. Sorry for the inconvenience.

**Comment by dpallotti** [comment]
> Something weird happened in this pull request. As you said, you didn't see a comment. Well, multiple comments are only visible to me, but they are not pending. I can see they don't show up if I open this page in incognito mode. Some GitHub bug. I added them again. Sorry for the inconvenience.

Actually I think they are only visible in the "Files changed" view. But when expanding the "# hidden conversations" in the main page, they don't show up. So use the "Files changed" view I guess.

File: src/WindowsUtils/DllInjection.cpp [hunk 32]
```diff
@@ -0,0 +1,278 @@
...
+
+      uint32_t function_rva = 0;
+      OUTCOME_TRY(
+          RemoteRead(handle, address_of_functions + ordinal * sizeof(function_rva), function_rva));
+      return module_base + function_rva;
+    }
+  }
+
+  return ErrorMessage(
+      absl::StrFormat("Did not find function \"%s\" in module \"%u\"", function_name, module_base));
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:275]: Regarding `\"%u\"`, did you want to print the module or the address?
- if the module, use `\"%s\"` and `module_name`;
- otherwise, use `%#x` and remove quotes.

File: src/WindowsUtils/DllInjectionTest.cpp [hunk 33]
```diff
@@ -0,0 +1,59 @@
...
+
+[[nodiscard]] std::filesystem::path GetTestDllPath() {
+  return orbit_test::GetTestdataDir() / "libtest.dll";
+}
+
+[[nodiscard]] std::filesystem::path GetNonExistentDllPath() { return "Z:/non_existent_dll.dll"; }
+
+}  // namespace
+
+TEST(Injection, InjectDllInCurrentProcess) {
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjectionTest.cpp:29]: Change `TEST(Injection` to `TEST(DllInjection` in all these tests.

File: src/WindowsUtils/DllInjection.cpp (hunk 18 above)
Comment (pierricgimmig) [reply src/WindowsUtils/DllInjection.cpp:50]: Yeah, I implemented your suggestion, it saved a few lines of code. The drawback is that we can't fill an existing object without a copy, but we don't have that use case here. Worse case we can add the previous implementation.

File: src/WindowsUtils/CMakeLists.txt (hunk 15 above)
Comment (pierricgimmig) [reply src/WindowsUtils/CMakeLists.txt]: Done.

File: src/WindowsUtils/DllInjection.cpp (hunk 19 above)
Comment (pierricgimmig) [reply src/WindowsUtils/DllInjection.cpp:56]: Changed the one above.

File: src/WindowsUtils/include/WindowsUtils/Injection.h (hunk 31 above)
Comment (pierricgimmig) [reply src/WindowsUtils/include/WindowsUtils/Injection.h]: Not sure what happened there, removed the file.

File: src/WindowsUtils/DllInjection.cpp (hunk 32 above)
Comment (pierricgimmig) [reply src/WindowsUtils/DllInjection.cpp]: Used module_name.

**Comment by pierricgimmig** [comment]
@dpallotti thanks for coping with the Github mishap. I addressed the remaining comments. Cheers.

File: src/WindowsUtils/include/WindowsUtils/DllInjection.h [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+#include <vector>
+
+#include "OrbitBase/Result.h"
+
+namespace orbit_windows_utils {
+
+// Injects a dll in a remote process identified by "pid".
+[[nodiscard]] ErrorMessageOr<void> InjectDll(uint32_t pid, std::filesystem::path dll_path);
+
+// Create a thread in a remote process and call specified function. The "parameter" byte buffer is
```
Comment (dpallotti) [inline src/WindowsUtils/include/WindowsUtils/DllInjection.h:19]:
```suggestion
// Create a thread in a remote process and call the specified function. The "parameter" byte buffer is
```

File: src/WindowsUtils/DllInjection.cpp [hunk 35]
```diff
@@ -0,0 +1,260 @@
...
+
+namespace {
+
+ErrorMessageOr<uint64_t> RemoteWrite(HANDLE process_handle, const std::vector<uint8_t>& buffer) {
+  // Allocate memory in target process.
+  LPVOID base_address = VirtualAllocEx(process_handle, /*lpAddress=*/0, buffer.size(),
+                                       MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
+  if (base_address == nullptr) {
+    return ErrorMessage(
+        absl::StrFormat("VirtualAllocEx failed with error %s", orbit_base::GetLastErrorAsString()));
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:31]: Should this then also be "Error calling VirtualAllocEx: %s"?

File: src/WindowsUtils/DllInjection.cpp [hunk 36]
```diff
@@ -0,0 +1,260 @@
...
+  if (!parameter.empty()) {
+    OUTCOME_TRY(uint64_t address, RemoteWrite(handle, parameter));
+    parameter_address = address;
+  }
+
+  if (!::CreateRemoteThread(handle, /*lpThreadAttributes=*/0, /*dwStackSize=*/0,
+                            absl::bit_cast<LPTHREAD_START_ROUTINE>(function_address),
+                            absl::bit_cast<LPVOID>(parameter_address),
+                            /*dwCreationFlags=*/0, /*lpThreadId=*/0)) {
+    return ErrorMessage(absl::StrFormat("CreateRemoteThread failed with error %s",
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjection.cpp:170]: And this "Error calling CreateRemoteThread: %s"?

File: src/WindowsUtils/DllInjectionTest.cpp [hunk 37]
```diff
@@ -0,0 +1,60 @@
...
+
+[[nodiscard]] std::filesystem::path GetNonExistentDllPath() { return "Z:/non_existent_dll.dll"; }
+
+}  // namespace
+
+TEST(DllInjection, InjectDllInCurrentProcess) {
+  // Injection.
+  uint32_t pid = orbit_base::GetCurrentProcessId();
+  ErrorMessageOr<void> result = InjectDll(pid, GetTestDllPath());
+  if (result.has_error()) ERROR("Running InjectDllInCurrentProcess: %s", result.error().message());
```
Comment (dpallotti) [inline src/WindowsUtils/DllInjectionTest.cpp:36]: I think you don't need this line as I think that we print the error anyway if any with `HasNoError`.

**Review by dpallotti — APPROVED** [review]
Looks good!
